core/src/main/scala/zio/logging/Logging.scala

```diff
@@ -0,0 +1,39 @@
+package zio.logging
+
+import zio.{ Cause, FiberRef, UIO, ZIO }
+
+trait AbstractLogging[Message] {
+  def logging: AbstractLogging.Service[Any, Message]
+}
+
+object AbstractLogging {
+
+  trait Service[-R, Message] {
+    def context: LoggingContext[Message]
+    def trace(message: Message): ZIO[R, Nothing, Unit]
+    def debug(message: Message): ZIO[R, Nothing, Unit]
+    def info(message: Message): ZIO[R, Nothing, Unit]
+    def warning(message: Message): ZIO[R, Nothing, Unit]
+    def error(message: Message): ZIO[R, Nothing, Unit]
+    def error(message: Message, cause: Cause[Any]): ZIO[R, Nothing, Unit]
+  }
+
+  final class LoggingContext[Message] private (ctx: FiberRef[Map[String, Message]]) {
```

Comment:
I think the context entries are not typed `Message`.  In Logstage, the context map is a `Map[String, Any]`, whereas `Message` is a structured log entry case class.

Maybe we need another generic type `Context` (which `trait Logging` can probably also fix to `String`)?

Reply:
my thinking is that those items from context you should be able to put in log statement as placeholders so it need to be align with ```Message```

Reply:
Why have a separate type for `LoggingContext`? i.e. why not just inline the `FiberRef` into the `Service`?

Reply:
If we need context to support values of any type, then we should not use `Any`. Instead, we should use something similar to ZIO Test `TestAnnotation`, which allows typed entries.

The question is:

 - Should we fix the context values to some type, like String or Message, or a new type parameter `Context`?
 - Or should we allow them to vary by key, like `TestAnnotationMap` / `TestAnnotation` in ZIO Test, in which case, we need a slightly different design here?

Reply:
> Why have a separate type for LoggingContext? i.e. why not just inline the FiberRef into the Service?

I believe that dealing with FiberRef it maybe not flexible enough and in cases that you need to even add something maybe a little bit cumbersome.

> Should we fix the context values to some type, like String or Message, or a new type parameter Context? 

what is a use case to add something to context in different type than ```Message```? those variables will end up in log file so they need to be in the same type as ```Message```  otherwise someone will need to provide those ugly converters. 

> Or should we allow them to vary by key, like TestAnnotationMap / TestAnnotation in ZIO Test, in which case, we need a slightly different design here?

I like that idea :)

Reply:
Here is a design:

```scala
trait ContextMap {
  def get[V](key: ContextKey[V]): FiberRef[V]
}
class ContextKey[V] private (
  val identifier: String,
  val initial: V,
  val combine: (V, V) => V,
  private val classTag: ClassTag[V]
) { ... }
```

Then you can have:

```scala
def addContext[V](contextKey: ContextKey[V], v: V): ZIO[R, Nothing, Unit]
def getContext[V](contextKey: ContextKey[V]): ZIO[R, Nothing, Unit]
```

Then perhaps last thing is `span` like operator which can aggregate context:

```scala
def span[R, E, A](name: Message)(zio: ZIO[R, E, A]): ZIO[R, E, A]
```

So you could do:

```scala
span("database access") {
   ... all database access ...
}
```

I'm not 100% sure about `span` but the contextual methods seem nice / useful.

Reply:
I will create ticket for ```span``` because that might be good idea. 

I replaced ```LoggingContext``` with ```ContextMap``` let me know what do you think about that.